agent/src/capability/baseconfigstore.cpp

```diff
@@ -23,173 +23,201 @@ namespace softwarecontainer {
 {
+    std::string errorMessage;
+
     if (!isJsonFile(filePath)) {
-        log_debug() << "File is not a json file: " << filePath;
-        return ReturnCode::FAILURE;
+        errorMessage = "File is not a json file";
+        log_debug() << errorMessage << ": " << filePath;
+        throw ServiceManifestPathError(errorMessage);
     }
 
     json_error_t error;
     json_t *fileroot = json_load_file(filePath.c_str(), 0, &error);
 
     if (nullptr == fileroot) {
-        log_error() << "Could not parse Service Manifest: "
-                    << filePath << ":" << error.line <<" : " << error.text;
-        return ReturnCode::FAILURE;
+        errorMessage = "Could not parse Service Manifest";
+        log_error() << errorMessage << ": " << filePath << ":" << error.line <<" : " << error.text;
+        throw ServiceManifestPathError(errorMessage);
     } else if (!json_is_object(fileroot)) {
-        log_error() << "Configuration is not a json object: \n"
-                    << filePath << ":" << error.line <<" : " << error.text;
-        return ReturnCode::FAILURE;
+        errorMessage = "Configuration is not a json object";
+        log_error() << errorMessage << ": " << filePath << ":" << error.line <<" : " << error.text;
+        throw ServiceManifestPathError("Service Manifest parse error - " + errorMessage);
     }
 
     json_t *capabilities = json_object_get(fileroot, "capabilities");
     if (nullptr == capabilities) {
-        log_error() << "Could not parse Capabilities";
-        return ReturnCode::FAILURE;
+        errorMessage = "Could not parse Capabilities";
+        log_error() << errorMessage;
+        throw CapabilityParseError(errorMessage);
     }
     if (!json_is_array(capabilities)) {
-        log_error() << "Capabilities is not an array";
-        return ReturnCode::FAILURE;
+        errorMessage = "Capabilities is not an array";
+        log_error() << errorMessage;
+        throw CapabilityParseError(errorMessage);
     }
 
     // Can't use json_decref on fileroot, it removes objects too early
-
-    return parseCapabilities(capabilities);
+    try {
+        parseCapabilities(capabilities);
+    } catch(CapabilityParseError &error) {
+        throw;
+    }
 }
 
-ReturnCode BaseConfigStore::parseCapabilities(json_t *capabilities)
+void BaseConfigStore::parseCapabilities(json_t *capabilities)
 {
     size_t i;
     json_t *capability;
+    std::string errorMessage;
 
     log_debug() << "Size of capabilities is " << std::to_string(json_array_size(capabilities));
 
     json_array_foreach(capabilities, i, capability) {
         if (!json_is_object(capability)) {
-            log_error() << "Capability is not a json object";
-            return ReturnCode::FAILURE;
+            errorMessage = "Capability is not a json object";
+            log_error() << errorMessage;
+            throw CapabilityParseError("Capability parse error - " + errorMessage);
```

Comment:
We already know this is a capability parse error, the type says that. This goes for the rest of the occurrences in this file as well.